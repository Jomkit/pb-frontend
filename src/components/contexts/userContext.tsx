import { createContext } from "react";
import { IUser } from "../../types";

const userContext = createContext({} as IUser);

export default userContext;