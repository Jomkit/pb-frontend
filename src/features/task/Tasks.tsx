import { useContext, useEffect, useState } from 'react';
import Frame from '../../components/Frame'
import Api from '../../api';
import TaskList from './TaskList';
import userContext from '../../components/contexts/userContext';
import CustomDropdown from '../../components/CustomDropdown';
import { IProject, ITask } from '../../types';
import useLocalStorage from '../../components/hooks/useLocalStorage';

const Tasks = () => {
  
  const [currProjectId, setCurrProjectId] = useLocalStorage("currProjectId");
  const [currProjectName, setCurrProjectName] = useState("No project selected");
  const { id } = useContext(userContext);

  // get tasks
  const [tasks, setTasks] = useState([] as ITask[]);
  const [loadingTasks, setLoadingTasks] = useState(true);
  const [projects, setProjects] = useState([] as IProject[]);

  useEffect(() => {
    const getUserProjects = async () => {
      const res = await Api.getUserProjects(id!);
      console.log("res in api", res);
      
      setProjects(res);
    }
    if(id){
      getUserProjects();
    }
    
  },[id]);
  
  useEffect(() => {
    const findTasksFromProject = async () => {
      const results = await Api.findProjectTasks(currProjectId as string);
      if(results.length !== 0){
        setTasks(results);
      }
      setLoadingTasks(false);
    }
    
    if(id){
      setLoadingTasks(true);
      
      let currProject;
      
      if(projects.length === 0) return;
      currProject = projects.find((proj) => proj.id === currProjectId);
      setCurrProjectName(currProject!.name);
      
      findTasksFromProject();
    }
  }, [currProjectId, projects]);

  const changeProject = (project: IProject) => {
    setCurrProjectId(project.id);
  }

  return (
    <Frame bgColor='bg-orange-400' title='Tasks'>
      {loadingTasks ? <p>Loading...</p> : 
      <>
        <CustomDropdown 
          label={"Select Project"} 
          items={projects} 
          handleClick={changeProject}
        />
        {projects.length > 1 && <h2>Current Project: {currProjectName}</h2>}
        <TaskList loading={loadingTasks} tasks={tasks} />
      </>
      }
    </Frame>
  )
}

export default Tasks