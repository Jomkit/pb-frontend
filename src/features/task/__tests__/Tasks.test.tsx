import { render } from "@testing-library/react"
import Tasks from "../Tasks"
import { MemoryRouter } from "react-router-dom";
import userContext from "../../../components/contexts/userContext";

const user = {
    id: 1,
    username: "testUser1",
    firstName: "Test",
    lastName: "User",
    email: "test@email.com",
}

beforeEach(() => {
    vi.mock('../../../api');
    vi.mock('../../../components/hooks/useLocalStorage', () => ({
        default: () => ["1", vi.fn()]
    }));
})
afterEach(() => {
    vi.restoreAllMocks();
})

it("should render without crashing", () => {
    render(<Tasks />);
})

it("should match snapshot of loading view", () => {
    const { asFragment } = render(<Tasks />);
    expect(asFragment()).toMatchSnapshot();
})

it("should match snapshot of loaded view", async () => {
    const { asFragment } = render(
        <MemoryRouter>
            <userContext.Provider value={user}>
                <Tasks />
            </userContext.Provider>
        </MemoryRouter>
    )
    
    await new Promise(resolve => setTimeout(resolve, 100));
    expect(asFragment()).toMatchSnapshot();
})