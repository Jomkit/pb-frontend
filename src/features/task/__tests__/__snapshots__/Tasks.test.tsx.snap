// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot of loaded view 1`] = `
<DocumentFragment>
  <section
    class="text-center items-start min-h-screen w-screen p-4 bg-orange-400"
  >
    <h1>
      Tasks
    </h1>
    <div
      class="p-10 justify-center"
    >
      <div
        class="simple-border rounded w-fit bg-slate-500 h-6"
      >
        <button
          aria-expanded="false"
          aria-haspopup="menu"
          class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 focus:outline-none border border-transparent bg-cyan-700 text-white focus:ring-4 focus:ring-cyan-300 enabled:hover:bg-cyan-800 dark:bg-cyan-600 dark:focus:ring-cyan-800 dark:enabled:hover:bg-cyan-700 rounded-lg w-fit"
          data-testid="flowbite-dropdown-target"
          id=":r1:"
          type="button"
        >
          <span
            class="flex items-stretch transition-all duration-200 rounded-md px-4 py-2 text-sm"
          >
            Select Project
            <svg
              aria-hidden="true"
              class="ml-2 h-4 w-4"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 9l-7 7-7-7"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </span>
        </button>
      </div>
      <h2>
        Current Project: testProject 1
      </h2>
      <div
        class="w-3/4 m-auto"
      >
        <div
          class="text-start"
        >
          <button
            class="btn bg-green-400 border-solid border-2 border-black"
          >
            New Task
          </button>
        </div>
        <div
          class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
        >
          <div
            class="text-center"
          >
            <h2>
              testTask 1
            </h2>
            <h4>
              Project id: 1
            </h4>
          </div>
          <a
            href="/tasks/1"
          >
            <p
              class="btn bg-blue-500 text-white"
            >
              Click for more info
            </p>
          </a>
          <div
            class="m-2"
          >
            <div
              class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
            >
              <div
                class="text-center"
              >
                <h2 />
                <h4 />
              </div>
              <h2
                class="text-start"
              >
                Timer
              </h2>
              <div
                class="flex justify-between"
              >
                <div
                  class="text-center"
                >
                  <h3
                    class="text-2xl bg-slate-200 w-fit m-auto px-4 py-1 rounded border-2 border-slate-500"
                  >
                    25:00
                  </h3>
                </div>
                <div
                  id="set-timer"
                >
                  <button
                    class="btn bg-green-400"
                    id="start-btn"
                  >
                    <i
                      class="fa-solid fa-play"
                    />
                  </button>
                  <button
                    class="btn bg-red-500"
                    id="restart-btn"
                  >
                    <i
                      class="fa-solid fa-rotate-right"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
        >
          <div
            class="text-center"
          >
            <h2>
              testTask 2
            </h2>
            <h4>
              Project id: 1
            </h4>
          </div>
          <a
            href="/tasks/2"
          >
            <p
              class="btn bg-blue-500 text-white"
            >
              Click for more info
            </p>
          </a>
          <div
            class="m-2"
          >
            <div
              class="w-full mx-auto mb-2 p-6 bg-white border-solid border-2 border-black rounded-lg divide-y"
            >
              <div
                class="text-center"
              >
                <h2 />
                <h4 />
              </div>
              <h2
                class="text-start"
              >
                Timer
              </h2>
              <div
                class="flex justify-between"
              >
                <div
                  class="text-center"
                >
                  <h3
                    class="text-2xl bg-slate-200 w-fit m-auto px-4 py-1 rounded border-2 border-slate-500"
                  >
                    25:00
                  </h3>
                </div>
                <div
                  id="set-timer"
                >
                  <button
                    class="btn bg-green-400"
                    id="start-btn"
                  >
                    <i
                      class="fa-solid fa-play"
                    />
                  </button>
                  <button
                    class="btn bg-red-500"
                    id="restart-btn"
                  >
                    <i
                      class="fa-solid fa-rotate-right"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`should match snapshot of loading view 1`] = `
<DocumentFragment>
  <section
    class="text-center items-start min-h-screen w-screen p-4 bg-orange-400"
  >
    <h1>
      Tasks
    </h1>
    <div
      class="p-10 justify-center"
    >
      <p>
        Loading...
      </p>
    </div>
  </section>
</DocumentFragment>
`;
